### PR TITLE
Change charms to use juju-http{s}-proxy model settings

### PR DESCRIPTION
**Pull request by VariableDeclared**
Changed charm to use juju-http{s}-proxy-settings when these keys are defined.

[LP#1831712](https://bugs.launchpad.net/charm-kubernetes-worker/+bug/1831712)

### CODE REVIEW

**Comment by VariableDeclared** [comment]
~~Pending Jenkins Tests
[#8](https://github.com/charmed-kubernetes/charm-containerd/pull/8)
[#138](https://github.com/juju-solutions/layer-docker/pull/138)~~

**Comment by VariableDeclared** [comment]
Added Jenkins Tests; [#447](https://github.com/charmed-kubernetes/jenkins/pull/447/files)

File: reactive/docker.py [hunk 1]
```diff
@@ -919,6 +921,18 @@ def recycle_daemon():
...
     """
+
+    # If juju environment variables are defined, take precedent
+    # over config.yaml.
+    # See: https://github.com/dshcherb/charm-helpers/blob/eba3742de6a7023f22778ba58fbbb0ac212d2ea6/charmhelpers/core/hookenv.py#L1455
+    # &: https://bugs.launchpad.net/charm-layer-docker/+bug/1831712
+    environment_config = hookenv.env_proxy_settings()
+    configuration = config()
+    if environment_config is not None:
+        configuration.update(environment_config)
```
Comment (Cynerva) [inline reactive/docker.py]: The [Config object](https://github.com/juju/charm-helpers/blob/b31408ece9f39b0bc527d785acd9ab492c14ff39/charmhelpers/core/hookenv.py#L278) returned by `config()` should be considered read-only. If the charm writes to it like this, then future hooks will get incorrect results from calls to `config.previous()`, and that will cause `config.changed.http_proxy`, `config.changed.https_proxy`, and `config.changed.no_proxy` to be set on every hook execution. That will ultimately lead to `recycle_daemon` being called on every hook.

**Comment by Cynerva** [comment]
Does this support CIDR notation in juju-no-proxy? From the [Juju docs](https://docs.jujucharms.com/offline-mode-strategies):

> One big benefit of using these finely-scoped settings is that juju-no-proxy can contain subnets (in CIDR notation) whereas its legacy counterpart cannot.

Docker most likely does not understand CIDR notation in the NO_PROXY env var (most apps don't), meaning we'd have to expand it into individual IPs somewhere along the way. I'm looking at the definition of [env_proxy_settings](https://github.com/juju/charm-helpers/blob/b31408ece9f39b0bc527d785acd9ab492c14ff39/charmhelpers/core/hookenv.py#L1424) and it doesn't look like it expands them, so that leaves it as the responsibility of the caller to do so.

**Comment by VariableDeclared** [comment]
Refactored and created another pull request on [container-runtime-common](https://github.com/charmed-kubernetes/layer-container-runtime-common/pull/4)

[main logic is here](https://github.com/charmed-kubernetes/layer-container-runtime-common/pull/4/files#diff-18083fbfdadac0ca35bd02a0aebda2c4R38)

**Comment by VariableDeclared** [comment]
Changing to WIP as I just need to run the most recent changes through.

**Comment by VariableDeclared** [comment]
Refactored and created another pull request on [container-runtime-common](https://github.com/charmed-kubernetes/layer-container-runtime-common/pull/4)

[main logic is here](https://github.com/charmed-kubernetes/layer-container-runtime-common/pull/4/files#diff-18083fbfdadac0ca35bd02a0aebda2c4R38)

File: reactive/docker.py [hunk 2]
```diff
@@ -180,7 +185,7 @@ def install():
         hookenv.log('Unknown runtime {}'.format(runtime))
         return False
 
-    validate_config()
+    validate_config(config())
```
Comment (Cynerva) [inline reactive/docker.py]: I think we want to use `check_for_juju_https_proxy` here, too. As it is now, it looks like this charm will initially configure and start the docker service without including the juju-proxy vars, until a future `recycle_daemon` corrects it.

File: README.md [hunk 3]
```diff
@@ -43,6 +43,8 @@ principal charm.
 See [config.yaml](config.yaml) for
 list of configuration options.
 
+> Note: Setting HTTP proxy values will be overriden if `juju-http-proxy` or `juju-https-proxy` settings are set on the model.
```
Comment (Cynerva) [inline README.md]: This is no longer true, correct?

File: README.md [hunk 4]
```diff
@@ -52,20 +54,24 @@ and logging.
...
   for bugs or problems with the Docker software.
   - The [charm-docker](https://github.com/juju-solutions/charm-docker) is
   the GitHub repository that contains the reactive code to build this Charm.
   - Check the charm-docker
   [issue-tracker](https://github.com/juju-solutions/charm-docker/issues) for
   bugs or problems related to the Charm.
+
+# Testing
+
+When upgrading this charm, the [`docker_runtime_version`](https://github.com/VariableDeclared/jenkins/blob/3bf2cabc7185568e6a80a137a7f083c428048683/jobs/validate-juju-https-envs.yaml#L22) of the `validate-https-proxy-envs` key needs to be updated with the new containerd version.
```
Comment (Cynerva) [inline README.md]: I suggest removing this from the README and removing the need to hard-code charm revs in the test.

More detailed comment: https://github.com/charmed-kubernetes/charm-containerd/pull/8#discussion_r305098886

File: README.md (hunk 3 above)
Comment (VariableDeclared) [reply README.md]: Cheers again!

File: reactive/docker.py (hunk 2 above)
Comment (VariableDeclared) [reply reactive/docker.py]: Makes sense

**Comment by Cynerva** [comment]
Thanks, LGTM :+1:

**Pull request merged**